### PR TITLE
[clang-format] Fix crashes when the macro expansion is empty

### LINKED ISSUE #119258: Empty macro crashes clang-format
```sh
$ clang-format --version

clang-format version 18.1.8 (Fedora 18.1.8-1.fc40)
```

```sh
$ echo "Macros: ['A']" > .clang-format
$ echo A | clang-format 

PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: clang-format
 #0 0x00007f4d8ee044fa llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/lib64/libLLVM.so.18.1+0x6044fa)
 #1 0x00007f4d8ee01ce4 llvm::sys::RunSignalHandlers() (/lib64/libLLVM.so.18.1+0x601ce4)
 #2 0x00007f4d8ee04c5b (/lib64/libLLVM.so.18.1+0x604c5b)
 #3 0x00007f4d8e24fd00 __restore_rt (/lib64/libc.so.6+0x40d00)
 #4 0x00007f4d98a31ee6 (/lib64/libclang-cpp.so.18.1+0x3031ee6)
 #5 0x00007f4d98a312f6 clang::format::TokenAnalyzer::process(bool) (/lib64/libclang-cpp.so.18.1+0x30312f6)
 #6 0x00007f4d98a06f2c (/lib64/libclang-cpp.so.18.1+0x3006f2c)
 #7 0x00007f4d989f072f clang::format::internal::reformat(clang::format::FormatStyle const&, llvm::StringRef, llvm::ArrayRef<clang::tooling::Range>, unsigned int, unsigned int, unsigned int, llvm::StringRef, clang::format::FormattingAttemptStatus*) (/lib64/libclang-cpp.so.18.1+0x2ff072f)
 #8 0x00007f4d989f1325 clang::format::reformat(clang::format::FormatStyle const&, llvm::StringRef, llvm::ArrayRef<clang::tooling::Range>, llvm::StringRef, clang::format::FormattingAttemptStatus*) (/lib64/libclang-cpp.so.18.1+0x2ff1325)
 #9 0x000055747ba75b41 (/usr/bin/clang-format+0xbb41)
#10 0x000055747ba735a4 (/usr/bin/clang-format+0x95a4)
#11 0x00007f4d8e239088 __libc_start_call_main (/lib64/libc.so.6+0x2a088)
#12 0x00007f4d8e23914b __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x2a14b)
#13 0x000055747ba726d5 (/usr/bin/clang-format+0x86d5)
fish: Process 6563, 'clang-format' from job 1, 'echo A | clang-format' terminated by signal SIGSEGV (Address boundary error)
```

### PR DESCRIPTION
**Pull request by amane-ame**
An empty expansion should be valid, like `echo 'A()' | clang-format --style='{Macros: [A(x)=x]}'`.

Fixes #119258.

### CODE REVIEW

**Comment by owenca** [comment]
Please add unit tests.

**Comment by owenca** [comment]
You should add unit tests in `clang/unittests/Format/`.

File: clang/unittests/Format/MacroCallReconstructorTest.cpp [hunk 1]
```diff
@@ -217,6 +217,19 @@ TEST_F(MacroCallReconstructorTest, Identifier) {
   EXPECT_THAT(std::move(Unexp).takeResult(), matchesLine(line(U.consume("X"))));
 }
 
+TEST_F(MacroCallReconstructorTest, EmptyExpansion) {
+  auto Macros = createExpander({"A(x)=y"});
```
Comment (owenca) [inline clang/unittests/Format/MacroCallReconstructorTest.cpp:221]: Changing `y` to `x` causes an assertion failure.

File: clang/lib/Format/MacroExpander.cpp [hunk 2]
```diff
@@ -233,6 +233,10 @@ MacroExpander::expand(FormatToken *ID,
   if (Result.size() > 1) {
     ++Result[0]->MacroCtx->StartOfExpansion;
     ++Result[Result.size() - 2]->MacroCtx->EndOfExpansion;
+  } else {
+    // If the macro expansion is empty, mark the start and end
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/MacroExpander.cpp:237]:
```suggestion
    // If the macro expansion is empty, mark the start and end.
```

**Review by owenca — COMMENTED** [review]
Please add a test case for https://github.com/llvm/llvm-project/issues/119258#issue-2728126922.

File: clang/test/Format/empty-macro.cpp [hunk 3]
```diff
@@ -0,0 +1,5 @@
+// RUN: grep -Ev "// *[A-Z-]+:" %s | clang-format -style="{Macros: [A(x)=x]}" \
+// RUN:   | FileCheck -strict-whitespace %s
+
+// CHECK: A()
+A()
```
Comment (owenca) [inline clang/test/Format/empty-macro.cpp:5]: Don't add a lit test. Instead, add a unittest similar to the  one you already added there.

Comment (amane-ame) [reply clang/test/Format/empty-macro.cpp:5]: Sorry for misunderstood.

File: clang/unittests/Format/MacroCallReconstructorTest.cpp [hunk 4]
```diff
@@ -217,6 +219,31 @@ TEST_F(MacroCallReconstructorTest, Identifier) {
   EXPECT_THAT(std::move(Unexp).takeResult(), matchesLine(line(U.consume("X"))));
 }
 
+TEST_F(MacroCallReconstructorTest, IdentifierObject) {
```
Comment (owenca) [inline clang/unittests/Format/MacroCallReconstructorTest.cpp:222]: How about`EmptyDefinition` instead of `IdentifierObject`?

**Review by owenca — APPROVED** [review]

**Pull request merged**